presto-native-execution/presto_cpp/main/http/filters/HttpEndpointLatencyFilter.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "presto_cpp/main/http/filters/HttpEndpointLatencyFilter.h"
+#include "velox/common/base/Exceptions.h"
+
+namespace facebook::presto::http::filters {
+
+HttpEndpointLatencyFilter::HttpEndpointLatencyFilter(
+    proxygen::RequestHandler* upstream,
+    const std::shared_ptr<std::unordered_map<
+        proxygen::HTTPMethod,
+        std::vector<std::unique_ptr<EndPoint>>>>& endpoints)
+    : Filter(upstream), endpoints_(endpoints) {}
+
+// static
+void HttpEndpointLatencyFilter::updateLatency(
+    const std::string& endpoint,
+    uint64_t latencyUs) {
+  metricMap().withWLock([&](std::unordered_map<std::string, EndPointMetrics>&
+                                map) {
+    auto itr = map.find(endpoint);
+    if (itr != map.end()) {
+      auto& metrics = itr->second;
+      metrics.maxLatencyUs = std::max(metrics.maxLatencyUs, latencyUs);
+      metrics.avgLatencyUs =
```

Comment:
not sure if you want to use double to track average latency to avoid int overflow. But I guess it is ok as we reset periodically.

Reply:
It should be good, uint64_t representing micro could represent max 5 billion hours. With 1mil QPS and 60s per request we can be fine for 5b/1m/60 which is 83min.